### PR TITLE
[ConvDiff] New semiimplicit solver

### PR DESCRIPTION
**Pull request by rubenzorrilla**
**Description**
This adds a new version of the old semi-implicit solver (BFECCSolver). The idea is to solve a pure diffusion problem (without convective term) and then do the convection with the BFECC algorithm.

This requires #7640 to be merged. Once this is added we can also merge #7537 which, among many other Python scripts, removes the legacy BFECC solver.

These new capabilities must be added to the GUI too.

Please mark the PR with appropriate tags: 
- Applications
- Feature

**Changelog**
- New BFECC convection-based thermal solver
- Testing with the already existent test cases in the application

### CODE REVIEW

File: applications/ConvectionDiffusionApplication/custom_elements/eulerian_diff.h [hunk 1]
```diff
@@ -1,12 +1,12 @@
...
+//      | (_| (_) | .` |\ V /___| |) | || _|| _|
 //       \___\___/|_|\_| \_/    |___/___|_| |_|  APPLICATION
 //
-//  License: BSD License
-//					 Kratos default license: kratos/license.txt
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
 //
-//  Main authors:  Riccardo Rossi
+//  Main authors:    Riccardo Rossi
```
Comment (RiccardoRossi) [inline applications/ConvectionDiffusionApplication/custom_elements/eulerian_diff.h:9]: @rubenzorrilla add yourself as author

File: applications/ConvectionDiffusionApplication/custom_elements/eulerian_diff.h [hunk 2]
```diff
@@ -221,18 +221,100 @@ class EulerianDiffusionElement
...
+        // Adding the diffusion
+        // Note 1: the diffusive term is computed using a Crank-Nicholson scheme
+        // Note 2: the gradients already include the corresponding Gauss point weight (no need to divide by TNumNodes)
+        const double aux_2 = conductivity * 0.5 * volume;
+        noalias(rRightHandSideVector) -= aux_2 * prod(prod(DN_DX, trans(DN_DX)), phi_convected + phi);
 
+        KRATOS_CATCH("Error in Eulerian diffusion element CalculateRightHandSide")
+    }
 
     void EquationIdVector(EquationIdVectorType& rResult, ProcessInfo& rCurrentProcessInfo) override
```
Comment (RiccardoRossi) [inline applications/ConvectionDiffusionApplication/custom_elements/eulerian_diff.h]: make this method const (and also the getdof)

File: applications/ConvectionDiffusionApplication/tests/cpp_tests/test_eulerian_diff.cpp [hunk 3]
```diff
@@ -163,8 +163,18 @@ namespace Testing
         expected_LHS(3, 2) = 0.0833333;
         expected_LHS(3, 3) = 0.25;
 
+        // Test CalculateLocalSystem
+        auto p_element = r_test_model_part.pGetElement(1);
+        Vector RHS = ZeroVector(4);
+        Matrix LHS = ZeroMatrix(4,4);
+        p_element->CalculateLocalSystem(LHS, RHS, r_test_model_part.GetProcessInfo());
         KRATOS_CHECK_VECTOR_NEAR(RHS, expected_RHS, 1.0e-4)
```
Comment (RiccardoRossi) [inline applications/ConvectionDiffusionApplication/tests/cpp_tests/test_eulerian_diff.cpp]: why so few decimals?

**Review by RiccardoRossi — COMMENTED** [review]
minor comments, otherwise good for me

**Review by RiccardoRossi — DISMISSED** [review]
well, approving "as is" however please take into account the comment about the const as otherwise it triggers a warning

File: applications/ConvectionDiffusionApplication/custom_elements/eulerian_diff.h (hunk 1 above)
Comment (rubenzorrilla) [reply applications/ConvectionDiffusionApplication/custom_elements/eulerian_diff.h:9]: Done.

File: applications/ConvectionDiffusionApplication/custom_elements/eulerian_diff.h (hunk 2 above)
Comment (rubenzorrilla) [reply applications/ConvectionDiffusionApplication/custom_elements/eulerian_diff.h]: This was already `const`. However, this wasn't the case for the getdof :+1:

File: applications/ConvectionDiffusionApplication/tests/cpp_tests/test_eulerian_diff.cpp (hunk 3 above)
Comment (rubenzorrilla) [reply applications/ConvectionDiffusionApplication/tests/cpp_tests/test_eulerian_diff.cpp]: This is what we used to have. I changed it to 10^-12 in all the tests.

**Review by RiccardoRossi — APPROVED** [review]

**Pull request merged**